sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLQuerySuite.scala

```diff
@@ -2348,4 +2348,36 @@ class SQLQuerySuite extends QueryTest with SQLTestUtils with TestHiveSingleton {
     }
   }
 
+  test("SPARK-25158: " +
+    "Executor accidentally exit because ScriptTransformationWriterThread throw Exception") {
+    withTempView("test") {
+      val defaultUncaughtExceptionHandler = Thread.getDefaultUncaughtExceptionHandler
+      try {
+        val uncaughtExceptionHandler = new TestUncaughtExceptionHandler
+        Thread.setDefaultUncaughtExceptionHandler(uncaughtExceptionHandler)
+
+        // Use a bad udf to generate failed inputs.
+        import org.apache.spark.sql.functions.udf
+        val badUDF = udf({x: Int =>
+          if (x < 1) x
+          else throw new RuntimeException("Failed to produce data.")
+          })
+        spark
+          .range(5)
+          .select(badUDF('id).as("a"))
+          .createOrReplaceTempView("test")
+        val scriptFilePath = getTestResourcePath("data")
+        val e = intercept[SparkException] {
+          sql(
+            s"""FROM test SELECT TRANSFORM(a)
+               |USING 'python $scriptFilePath/scripts/test_transform.py "\t"'
+             """.stripMargin).collect()
+        }
+        assert(e.getMessage.contains("Failed to produce data."))
```

Comment:
what's the result if we run this test without your fix and the `TestUncaughtExceptionHandler`?

Reply:
Keep `throw t`, `TestUncaughtExceptionHandler` will catch a `SparkException`.
Keep `throw t` and not  set  `TestUncaughtExceptionHandler` as DefaultUncaughtExceptionHandler,  this case always success but we can see log as follwing:
```
ERROR org.apache.spark.util.Utils: Uncaught exception in thread Thread-ScriptTransformation-Feed
```

Reply:
And if keep throw t and use `SparkUncaughtExceptionHandler` instead of `TestUncaughtExceptionHandler` , the case always success but we can see log as follwoing:

```
ERROR org.apache.spark.util.Utils: Uncaught exception in thread Thread-ScriptTransformation-Feed
ERROR org.apache.spark.util.SparkUncaughtExceptionHandler: Uncaught exception in thread Thread[Thread-ScriptTransformation-Feed,5,main]
WARN org.apache.spark.sql.hive.execution.SQLQuerySuite: Job 0 cancelled because SparkContext was shut down
Process finished with exit code 50
```

Reply:
> The test process finished with exit code 50 not 0

What's the return data of the query then?

Reply:
No return data... we intercept a `SparkException`